src/udp-echo-server.1
.\" ========================================================================
.\"    _   _ _ ____            ____          _____
.\"   | | | (_)  _ \ ___ _ __ / ___|___  _ _|_   _| __ __ _  ___ ___ _ __
.\"   | |_| | | |_) / _ \ '__| |   / _ \| '_ \| || '__/ _` |/ __/ _ \ '__|
.\"   |  _  | |  __/  __/ |  | |__| (_) | | | | || | | (_| | (_|  __/ |
.\"   |_| |_|_|_|   \___|_|   \____\___/|_| |_|_||_|  \__,_|\___\___|_|
.\"
.\"      ---  High-Performance Connectivity Tracer (HiPerConTracer)  ---
.\"                https://www.nntb.no/~dreibh/hipercontracer/
.\" ========================================================================
.\"
.\" High-Performance Connectivity Tracer (HiPerConTracer)
.\" Copyright (C) 2015-2026 by Thomas Dreibholz
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" Contact: dreibh@simula.no
.\"
.\" ###### Setup ############################################################
.Dd September 7, 2026
.Dt udp\-echo\-server 1
.Os udp\-echo\-server
.\" ###### Name #############################################################
.Sh NAME
.Nm udp\-echo\-server
.Nd Simple UDP Echo Server
.\" ###### Synopsis #########################################################
.Sh SYNOPSIS
.Nm udp\-echo\-server
.Op Fl L Ar level | Fl \-loglevel Ar level
.br
.Op Fl Z Ar on|off | Fl \-logcolor Ar on|off
.br
.Op Fl O Ar file | Fl \-logfile Ar file
.br
.Op Fl q | Fl \-quiet
.Op Fl w | Fl \-verbose
.br
.Op Fl U Ar user|uid | Fl \-user Ar user|uid
.br
.Op Fl A Ar address | Fl \-address Ar address
.br
.Op Fl P Ar port | Fl \-port Ar port
.br
.Op Fl f Ar port | Fl \-from\-port Ar port
.Op Fl t Ar port | Fl \-to\-port Ar port
.Nm udp\-echo\-server
.Op Fl h | Fl \-help
.Nm udp\-echo\-server
.Op Fl v | Fl \-version
.\" ###### Description ######################################################
.Sh DESCRIPTION
.Nm udp\-echo\-server
is a simple UDP Echo server. It can be used as endpoint for HiPerConTracer
UDP Ping/Traceroute.
Note, for security reasons, udp\-echo\-server only responds to source ports
>= 1024, to avoid using the server in attacks, like spoofing a packet from
another Echo server (port 7) to create a flooding loop. Also, packets
from the same port as the listening port are ignored.
.Pp
.\" ###### Arguments ########################################################
.Sh ARGUMENTS
The following arguments may be provided:
.Bl -tag -width indent
.It Fl L Ar level | Fl \-loglevel Ar level
Sets the minimum logging level to the provided value: 0=trace, 1=debug, 2=info, 3=warning, 4=error, 5=fatal.
.It Fl q | Fl \-quiet
Sets the minimum logging level to 3 (warning).
.It Fl w | Fl \-verbose
Sets the minimum logging level to 0 (trace).
.It Fl Z Ar on|off | Fl \-logcolor Ar on|off
Enables (on) or disables (off) the usage of ANSI color escape sequences to colorise the log output according to its level.
.It Fl O Ar file | Fl \-logfile Ar file
Sets a file name for appending the log output to. If a file is set, no log output is made to the console. Otherwise, all log output will be written to the console.
.It Fl U Ar user|uid | Fl \-user Ar user|uid
After startup, the server uses UID and GID of the given user (by name or GID).
.It Fl A Ar address | Fl \-address Ar address
Sets the local address (default: ::).
.It Fl P Ar port | Fl \-port Ar port
Sets the local port (default: 7).
.It Fl f Ar port | Fl \-from\-port Ar port
Sets the first local port of a range (default: 7).
.It Fl t Ar port | Fl \-to\-port Ar port
Sets the last local port of a range (default: 0; 0 uses the "from" port).
.It Fl h | Fl \-help
Prints command help.
.It Fl v | Fl \-version
Prints program version.
.El
.\" ###### Exit status ######################################################
.Sh EXIT STATUS
The
.Nm
tool exits with 0 on success, and >0 in case of an error.
.\" ###### Examples #########################################################
.Sh EXAMPLES
.Bl -tag -width indent
.It udp\-echo\-server
.It udp\-echo\-server \-\-address 192.168.1.1 \-\-port 7777
.It udp\-echo\-server \-\-address 3ffe::1:2:3 \-\-port 7777
.It udp\-echo\-server \-\-address :: \-\-user nornetpp
.It udp\-echo\-server \-\-address 3ffe::1:2:3 \-\-from\-port 10000 \-\-to\-port 10999
.It udp\-echo\-server \-\-version
.It udp\-echo\-server \-h
.El
.\" ###### See also #########################################################
.Sh SEE ALSO
.Xr hipercontracer 1
.Xr hpct-trigger 1
.\" ###### Notes ############################################################
.Sh NOTES
This program is part of HiPerConTracer. The latest version of HiPerConTracer can be found on the HiPerConTracer Homepage at
.Lk https://www.nntb.no/\(tidreibh/hipercontracer/ "HiPerConTracer Homepage" .
.\" ###### Authors ##########################################################
.Sh AUTHORS
Thomas Dreibholz,
.Lk https://www.nntb.no/\(tidreibh "Homepage"


ChangeLog
commit c2586671eedee0068fc623622fe59f058b1e40c2
Merge: b00a9968 1c3000d6
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:59:49 2026 +0200

    Merge branch 'dreibh/homebrew' of github.com:dreibh/hipercontracer into dreibh/homebrew

commit b00a9968e3a706ba6c4b79e7fd74d51753cc3ba1
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:58:41 2026 +0200

    Removed exit() call from library. DatabaseConfiguration::readConfiguration() returns false in case of a path error.

commit 1c3000d647dae82789040f95ff5a1b8b3853a65c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:55:16 2026 +0200

    Updated RPM spec file.

commit a2188f7125d634ff6e0d0ede00c527730536205e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:47:00 2026 +0200

    Using CMAKE_INSTALL_SYSCONFDIR for configuration file installation instead of hard-coding /etc.

commit 52ac9bc09bb4165923d0e3b2781497cab9d9f3fb
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:30:42 2026 +0200

    Improved Homebrow path handling in CMakeLists.txt.

commit 2be6cc58880f261804db093b06920e4f13522ecb
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:10:01 2026 +0200

    Some progress with the Homebrew port.

commit 21a740847032ad7e2b86da2571211a37e641d2bc
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 22:57:47 2026 +0200

    Some improvements.

commit f8a3f894156a3938dc44c7e3015ee4b50ab2142d
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 22:43:15 2026 +0200

    Initial homebrew packaging. Not working, yet!

commit 112761ee05490e780af98b8b80b29d4667c3ac86
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 17:07:58 2026 +0200

    Improved RPM spec file.

commit 4ab4e15001f36985d99663396bf93af46bd22056
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 16:57:57 2026 +0200

    Improved documentation.

commit 2f71eef68655e24b53083325c2bf7b6c569f3cd0
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 16:38:42 2026 +0200

    Setting capabilities (setcap cap_net_raw=ep) for Debian/Ubuntu and Fedora package.

commit 669311cc09b610fab6a00f170830bd9535d05980
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 16:37:25 2026 +0200

    Linux: Allow HiPerConTracer to run with capabilities (setcap cap_net_raw=ep) instead of startup with root permissions.

commit ef1049418431815e5674099d26b8d55f69770031
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 16:00:14 2026 +0200

    Another set of manpage improvements.

commit 3763e8cea0b34be3a7d0b76bc70b7e320f27f2f9
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 15:54:43 2026 +0200

    Further manpage improvements.

commit b0f30cc37b8ad936594ea882a6a80d07f05d9a9c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 15:37:42 2026 +0200

    Updated manpages.

commit f2ddd54b04e55f370bf96964937d7586b93abcad
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 15:37:23 2026 +0200

    Use positional parameters as destinations.

commit b701a523bb3e3177283a1254527dfb63f121b216
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 18:22:15 2026 +0200

    Updated APKBUILD file.

commit cc74111202cfbc62d05470f40a9f4fd29b1b0d1c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 18:07:12 2026 +0200

    New release hipercontracer-2.2.10.

commit e2318e7c008132ecf89cb3391c9906a92d3962a7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 15:10:01 2026 +0200

    Fix for Debian test 30-dbeaver-tools.

commit 813403b8077be2520460d15674cd634399ec11db
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 16:23:32 2026 +0200

    Updated APKBUILD file.

commit 5c3c467cccde4b0e0fb78868650085f2fd890c94
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 15:08:02 2026 +0200

    Improved example scripts.

commit bb7d3cbfe79dc06f3c4b2a5fc3687cbb2d5b1c80
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:57:44 2026 +0200

    Updated script.

commit 25c5684be31fec6f4c923e94461509b4c71552ae
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:54:06 2026 +0200

    Added test script.

commit bb15335db526f34f8322d27270d63064eed58e3f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:51:51 2026 +0200

    Added Ping example script as well.

commit f2cdf7327f20c5328b9b075839ce913bd83a56de
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:46:11 2026 +0200

    Further clean-ups.

commit b29a86abd30967b176eb5266098e2e16039d24e4
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:39:31 2026 +0200

    Further improvements of the examples, and some function name clean-ups.

commit 14334a09f1948fa48c0648274b11b13839ef9da9
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 12:53:25 2026 +0200

    Added another Python example.

commit 3269229394f112f1196d532d39992d2e344201a3
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 12:53:17 2026 +0200

    Improved Python library.

commit 3444d978211dfbab34ca19ced5eec21c350dd822
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 16:33:02 2026 +0200

    New release hipercontracer-2.2.9.

commit b9f4f55dab00295fc3c299777a4718b630170651
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 16:29:16 2026 +0200

    Version bump.

commit 1847ee70e030bc1c5fff9c9a0256b83af48a0ad2
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 15:28:13 2026 +0200

    Updated test scripts.

commit 269f102483fea833bfd2aae28e8b9036ab180e89
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 14:51:17 2026 +0200

    Manpage fix.

commit cebc281513d0e5ee95bae5eec470356485305c1a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 14:49:31 2026 +0200

    Improved manpages.

commit c3cc031b9c8a3363ab2ddbc2af3a028163a1ee72
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Aug 11 15:43:18 2026 +0200


src/udp-echo-server.cc
// ==========================================================================
//     _   _ _ ____            ____          _____
//    | | | (_)  _ \ ___ _ __ / ___|___  _ _|_   _| __ __ _  ___ ___ _ __
//    | |_| | | |_) / _ \ '__| |   / _ \| '_ \| || '__/ _` |/ __/ _ \ '__|
//    |  _  | |  __/  __/ |  | |__| (_) | | | | || | | (_| | (_|  __/ |
//    |_| |_|_|_|   \___|_|   \____\___/|_| |_|_||_|  \__,_|\___\___|_|
//
//       ---  High-Performance Connectivity Tracer (HiPerConTracer)  ---
//                 https://www.nntb.no/~dreibh/hipercontracer/
// ==========================================================================
//
// High-Performance Connectivity Tracer (HiPerConTracer)
// Copyright (C) 2015-2026 by Thomas Dreibholz
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// Contact: dreibh@simula.no

#include "package-version.h"

#include <iostream>
#include <boost/asio.hpp>
#include <boost/program_options.hpp>

#include "logger.h"
#include "tools.h"


class UDPEchoInstance
{
   public:
   UDPEchoInstance(boost::asio::io_context&        ioContext,
                   const boost::asio::ip::address& localAddress,
                   const uint16_t                  localPort);
   ~UDPEchoInstance();

   void handleMessage(const boost::system::error_code& errorCode,
                      std::size_t                      bytesReceived);

   boost::asio::io_context&             IOContext;
   const boost::asio::ip::udp::endpoint LocalEndpoint;
   boost::asio::ip::udp::socket         UDPSocket;
   boost::asio::ip::udp::endpoint       RemoteEndpoint;
   char                                 Buffer[65536];
};


// ###### Constructor #######################################################
UDPEchoInstance::UDPEchoInstance(boost::asio::io_context&        ioContext,
                                 const boost::asio::ip::address& localAddress,
                                 const uint16_t                  localPort)
   : IOContext(ioContext),
     LocalEndpoint(localAddress, localPort),
     UDPSocket(IOContext, (localAddress.is_v6() == true) ?
                              boost::asio::ip::udp::v6() :
                              boost::asio::ip::udp::v4())
{
   UDPSocket.bind(LocalEndpoint);
   UDPSocket.async_receive_from(boost::asio::buffer(Buffer, sizeof(Buffer)),
                                RemoteEndpoint,
                                std::bind(&UDPEchoInstance::handleMessage,
                                          this,
                                          std::placeholders::_1,
                                          std::placeholders::_2));
}


// ###### Destructor ########################################################
UDPEchoInstance::~UDPEchoInstance()
{
}


// ###### Handle incoming echo request ######################################
void UDPEchoInstance::handleMessage(const boost::system::error_code& errorCode,
                                    std::size_t                      bytesReceived)
{
   if(errorCode != boost::asio::error::operation_aborted) {
      // ====== Send response ===============================================
      if(bytesReceived > 0) {
         // Security:
         // * For security reasons, only respond if remote port >= 1024.
         //   Otherwise, an attacker could inject a message to cause
         //   an echo loop between two echo servers (port 7)!
         // * Also ignore requests from same port as local port!
         if( (RemoteEndpoint.port() >= 1024) &&
             (RemoteEndpoint.port() != LocalEndpoint.port()) ) {
            boost::system::error_code sendErrorCode;
            UDPSocket.send_to(boost::asio::buffer(Buffer, bytesReceived),
                              RemoteEndpoint, 0, sendErrorCode);
            if(sendErrorCode != boost::system::errc::success) {
               HPCT_LOG(error) << "Error sending " << bytesReceived << " bytes to "
                               << RemoteEndpoint << ": " << sendErrorCode.message();
            }
#if 0
            else {
               HPCT_LOG(trace) << "Replied " << bytesReceived << " bytes to "
                               << RemoteEndpoint;
            }
#endif
         }
         else {
            HPCT_LOG(warning) << "Ignoring request from " << RemoteEndpoint;
         }
      }

      // ====== Wait for next message =======================================
      UDPSocket.async_receive_from(boost::asio::buffer(Buffer, sizeof(Buffer)),
                                   RemoteEndpoint,
                                   std::bind(&UDPEchoInstance::handleMessage,
                                             this,
                                             std::placeholders::_1,
                                             std::placeholders::_2));
   }
}


// ###### Handle SIGINT #####################################################
static void signalHandler(boost::asio::io_context*         ioContext,
                          const boost::system::error_code& errorCode,
                          int                              signalNumber)
{
   std::cout << "\nGot signal " << signalNumber << "\n";
   ioContext->stop();
}



// ###### Main program ######################################################
int main(int argc, char *argv[])
{
   // ====== Initialize =====================================================
   unsigned int             logLevel;
   bool                     logColor;
   std::filesystem::path    logFile;
   std::string              user((getlogin() != nullptr) ? getlogin() : "0");
   std::string              localAddressString;
   boost::asio::ip::address localAddress;
   uint16_t                 localPortFrom;
   uint16_t                 localPortTo;

   boost::program_options::options_description commandLineOptions;
   commandLineOptions.add_options()
      ( "help,h",
           "Print help message" )
      ( "version,v",
           "Show program version" )

      ( "loglevel,L",
           boost::program_options::value<unsigned int>(&logLevel)->default_value(boost::log::trivial::severity_level::info),
           "Set logging level" )
      ( "logfile,O",
           boost::program_options::value<std::filesystem::path>(&logFile)->default_value(std::filesystem::path()),
           "Log file" )
      ( "logcolor,Z",
           boost::program_options::value<bool>(&logColor)->default_value(true),
           "Use ANSI color escape sequences for log output" )
      ( "verbose,w",
           boost::program_options::value<unsigned int>(&logLevel)->implicit_value(boost::log::trivial::severity_level::trace),
           "Verbose logging level" )
      ( "quiet,q",
           boost::program_options::value<unsigned int>(&logLevel)->implicit_value(boost::log::trivial::severity_level::warning),
           "Quiet logging level" )
      ( "user,U",
           boost::program_options::value<std::string>(&user),
           "User" )

      ( "address,A",
           boost::program_options::value<std::string>(&localAddressString)->default_value("::"),
           "Address" )
      ( "from-port,port,f,P",
           boost::program_options::value<uint16_t>(&localPortFrom)->default_value(7),
           "Port" )
      ( "to-port,t",
           boost::program_options::value<uint16_t>(&localPortTo)->default_value(0),
           "Port" )
      ;


   // ====== Handle command-line arguments ==================================
   boost::program_options::variables_map vm;
   try {
      boost::program_options::store(boost::program_options::command_line_parser(argc, argv).
                                       style(
                                          boost::program_options::command_line_style::style_t::unix_style
                                       ).
                                       options(commandLineOptions).
                                       run(), vm);
      boost::program_options::notify(vm);
   }
   catch(std::exception& e) {
      std::cerr << "ERROR: Bad parameter: " << e.what() << "\n";
      return 1;
   }

   if(localPortTo == 0) {
      localPortTo = localPortFrom;
   }
   if( (localPortTo < 1) || (localPortFrom > localPortTo) || (localPortTo > 65535) ) {
      std::cerr << "ERROR: Invalid port range " << localPortFrom << " - " << localPortTo << "\n";
      return 1;
   }
   const unsigned int ports = 1U + localPortTo - localPortFrom;

   try {
      localAddress = boost::asio::ip::make_address(localAddressString);
   }
   catch(std::exception& e) {
      std::cerr << "ERROR: Invalid address: " << e.what() << "\n";
      return 1;
   }

   if(vm.count("help")) {
       std::cerr << "Usage: " << argv[0] << " OPTIONS" << "\n"
                 << commandLineOptions;
       return 1;
   }
   else if(vm.count("version")) {
      std::cerr << "UDP Echo Server" << " " << HPCT_VERSION << "\n";
      return 0;
   }


   // ====== Initialize =====================================================
   initialiseLogger(logLevel, logColor,
                    (!logFile.empty()) ? logFile.string().c_str() : nullptr);
   const passwd* pw = getUser(user.c_str());
   if(pw == nullptr) {
      HPCT_LOG(fatal) << "Cannot find user \"" << user << "\"!";
      return 1;
   }

   // ------ Handle SIGINT --------------------------------------------------
   boost::asio::io_context ioContext;
   boost::asio::signal_set signals(ioContext, SIGINT);
   signals.async_wait(std::bind(&signalHandler, &ioContext,
                                std::placeholders::_1,
                                std::placeholders::_2));

   // ------ Create UDP instances -------------------------------------------
   UDPEchoInstance*        udpEchoInstance[ports];
   for(unsigned int p = 0; p  < ports; p++) {
      const uint16_t localPort = p + localPortFrom;
      try {
         udpEchoInstance[p] = new UDPEchoInstance(ioContext, localAddress, localPort);
      }
      catch(std::exception& e) {
         HPCT_LOG(fatal) << "Unable to bind UDP socket to source address "
                        << localAddress << ":" << localPort << ": " << e.what();
         return 1;
      }
   }
   if(ports > 1) {
      HPCT_LOG(info) << "Listening on UDP ports " << localPortFrom << " to " << localPortTo;
   }
   else {
      HPCT_LOG(info) << "Listening on UDP port " << localPortFrom;
   }

   // ====== Reduce privileges ==============================================
   if(reducePrivileges(pw) == false) {
      HPCT_LOG(fatal) << "Failed to reduce privileges!";
      return 1;
   }

   // ====== Main loop ======================================================
   ioContext.run();

   // ====== Clean up =======================================================
   for(unsigned int p = 0; p  < ports; p++) {
      delete udpEchoInstance[p];
      udpEchoInstance[p] = nullptr;
   }
   return 0;
}
